runtime/src/prioritization_fee_cache.rs

```diff
@@ -282,48 +264,52 @@ impl PrioritizationFeeCache {
     /// Internal function is invoked by worker thread to update slot's minimum prioritization fee,
     /// Cache lock contends here.
     fn update_cache(
-        cache: Arc<RwLock<LruCache<Slot, Arc<SlotPrioritizationFee>>>>,
-        slot: &Slot,
-        bank_id: &BankId,
+        unfinalized: &mut UnfinalizedPrioritizationFees,
+        slot: Slot,
+        bank_id: BankId,
         transaction_fee: u64,
-        writable_accounts: Arc<Vec<Pubkey>>,
-        metrics: Arc<PrioritizationFeeCacheMetrics>,
+        writable_accounts: Vec<Pubkey>,
+        metrics: &PrioritizationFeeCacheMetrics,
     ) {
-        let (slot_prioritization_fee, cache_lock_time) =
-            measure!(Self::get_prioritization_fee(cache, slot), "cache_lock_time");
-
         let (_, entry_update_time) = measure!(
             {
-                let mut block_prioritization_fee = slot_prioritization_fee
-                    .entry(*bank_id)
-                    .or_insert(PrioritizationFee::default());
-                block_prioritization_fee.update(transaction_fee, &writable_accounts)
+                let _ = unfinalized
+                    .entry(slot)
+                    .or_default()
+                    .entry(bank_id)
+                    .or_default()
+                    .update(transaction_fee, writable_accounts);
             },
             "entry_update_time"
         );
-        metrics.accumulate_total_cache_lock_elapsed_us(cache_lock_time.as_us());
         metrics.accumulate_total_entry_update_elapsed_us(entry_update_time.as_us());
         metrics.accumulate_successful_transaction_update_count(1);
     }
 
     fn finalize_slot(
-        cache: Arc<RwLock<LruCache<Slot, Arc<SlotPrioritizationFee>>>>,
-        slot: &Slot,
-        bank_id: &BankId,
-        metrics: Arc<PrioritizationFeeCacheMetrics>,
+        unfinalized: &mut UnfinalizedPrioritizationFees,
+        cache: &RwLock<LruCache<Slot, PrioritizationFee>>,
+        slot: Slot,
+        bank_id: BankId,
+        metrics: &PrioritizationFeeCacheMetrics,
     ) {
-        let (slot_prioritization_fee, cache_lock_time) =
-            measure!(Self::get_prioritization_fee(cache, slot), "cache_lock_time");
-
         // prune cache by evicting write account entry from prioritization fee if its fee is less
         // or equal to block's minimum transaction fee, because they are irrelevant in calculating
         // block minimum fee.
-        let (result, slot_finalize_time) = measure!(
+        let (slot_prioritization_fee, slot_finalize_time) = measure!(
             {
+                // remove unfinalized slots
+                *unfinalized = unfinalized
+                    .split_off(&slot.checked_sub(MAX_UNFINALIZED_SLOTS).unwrap_or_default());
+
+                let Some(mut slot_prioritization_fee) = unfinalized.remove(&slot) else {
+                    return;
```

Comment:
nit: Wouldn't normally reach here, perhaps warn log it, or report to metrics?

Reply:
`warn!` added :+1: